fix(datagrid): clicking select all checkbox focus loss

wkeese opened the pull request:
Closes #7348, and refs #5638.

Clicking the select-all checkbox triggered React to remove the parent `TH` and reinsert it into the document... causing the focus to be lost.

Avoided that by fixing spurious recreation of functions.

#### What did you change?

The problem was that React's VDOM diffing wasn't working well.  But to solve it, it wasn't just about adding `useCallback()` for the event handler callbacks.  I also needed to prevent spurious recreation of the `Heading` and `Cell` functions.

##### DatagridSelectAll.tsx

* Add `useCallback()` for `handleOnPageSelectAllChange()` and `handleSelectAllChange()`.

* Subsequently, move around variable declarations to obey rules of hooks, specifically that all the `useCallback()` need to happen before any `return` statement.

##### useSelectRows.tsx

* Create static `renderSelectAll()` and `renderSelectCell()` functions so that `Header` and `Cell` functions aren't recreated every render. I think this was the root issue.  See https://stackoverflow.com/questions/58507656/why-does-react-discard-the-entire-dom-subtree-and-recreate-it-from-scratch and https://legacy.reactjs.org/docs/reconciliation.html#elements-of-different-types.  But it's hard to tell because Datagrid is confused about the difference between components (that should start with a capital letter) and mere callbacks (that should start with a lowercase letter).

* Add `useCallback()` for `onSelectHandler()`.

* Move `onSelectHandler()` so that `onChange()` is not referenced before defined.

* Revert workaround from #6073 to refocus radio buttons after click, because this fixes the root cause of #5683.

* Use `aria-label` rather than deprecated `ariaLabel`.

#### How did you test and verify your work?

Storybook

wkeese commented:
CC @devadula-nandan.

devadula-nandan reviewed (dismissed):
@wkeese looks great

davidmenendez reviewed (changes requested):
thanks for submitting these changes. can you include or update the tests to account for this? seeing a few drops in code coverage. thanks!

wkeese commented:
> thanks for submitting these changes. can you include or update the tests to account for this? seeing a few drops in code coverage. thanks!

@davidmenendez - I'm pretty sure that's not true, and that it's just pointing out code paths that were _never_ covered.  It's just confused because I moved around the code.  Nonetheless, in the interest of merging this, I added more tests.

However, I'm not sure what's happening with [the codecov report](https://app.codecov.io/gh/carbon-design-system/ibm-products/pull/7385?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=carbon-design-system):

1. It says that in useSelectRows.ts, onRadioSelect() isn't being called but I tested in the debugger (running the 'Radio Select' test) and it is called.
2. It shows that I greatly improved testing in DatagridSelectAll.tsx.  The lines marked with ⚠ are all from the old code.
3. So why does this page claim that coverage went down?

anamikaanu96 reviewed (approved):
LGTM
